File: src/main/java/com/parmet/squashlambdas/cal/EventManager.java
package com.parmet.squashlambdas.cal;

import static com.google.common.base.Preconditions.checkNotNull;

import com.google.api.services.calendar.Calendar;
import com.google.api.services.calendar.model.Event;
import com.google.common.collect.Iterables;
import com.parmet.squashlambdas.activity.Activity;
import java.io.IOException;
import java.util.List;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class EventManager {
  private static final Logger log = LogManager.getLogger();

  private final Calendar calendar;
  private final String calendarId;

  public EventManager(Calendar calendar, String calendarId) {
    this.calendar = checkNotNull(calendar, "calendar");
    this.calendarId = checkNotNull(calendarId, "calendarId");
  }

  public void create(Activity activity) throws IOException {
    log.info("Creating schdulable {}", activity);
    calendar.events().insert(calendarId, activity.toEvent()).execute();
  }

  public void update(Activity activity) throws IOException {
    log.info("Updating activity {}", activity);

    List<Event> events = findEvents(activity);

    if (events.size() > 1) {
      log.info("Found too many events to update; deleting them all. {}", events);
      for (Event event : events) {
        calendar.events().delete(calendarId, event.getId()).execute();
      }
      create(activity);
    } else {
      Event event = Iterables.getOnlyElement(events);
      log.info("Found one event to update: {}", event);
      calendar.events().patch(calendarId, event.getId(), activity.toEvent()).execute();
    }
  }

  public void delete(Activity activity) throws IOException {
    log.info("Deleting activity {}", activity);
    for (Event event : findEvents(activity)) {
      log.info("Deleting event {}", event);
      calendar.events().delete(calendarId, event.getId()).execute();
    }
  }

  private List<Event> findEvents(Activity activity) throws IOException {
    return calendar.events().list(calendarId)
        .setQ(activity.searchString())
        .execute()
        .getItems();
  }
}

File: src/main/kotlin/com/parmet/squashlambdas/cal/EventManager.kt
package com.parmet.squashlambdas.cal

import com.google.api.services.calendar.Calendar
import com.google.common.collect.Iterables
import com.parmet.squashlambdas.activity.Activity
import mu.KotlinLogging

interface EventManager {
    fun create(activity: Activity)

    fun update(activity: Activity)

    fun delete(activity: Activity)
}

class EventManagerImpl(
    private val calendar: Calendar,
    private val calendarId: String
) : EventManager {
    private val logger = KotlinLogging.logger { }

    override fun create(activity: Activity) {
        logger.info { "Creating activity $activity" }
        calendar.events().insert(calendarId, activity.toEvent()).execute()
    }

    override fun update(activity: Activity) {
        logger.info { "Updating activity $activity" }

        val events = findEvents(activity)

        if (events.size > 1) {
            logger.info { "Found too many events to update; deleting them all. $events" }
            events.forEach {
                calendar.events().delete(calendarId, it.id).execute()
            }
            create(activity)
        } else {
            val event = Iterables.getOnlyElement(events)
            logger.info { "Found one event to update: $event" }
            calendar.events().patch(calendarId, event.id, activity.toEvent()).execute()
        }
    }

    override fun delete(activity: Activity) {
        logger.info { "Deleting activity $activity" }
        findEvents(activity).forEach {
            logger.info { "Deleting event $it" }
            calendar.events().delete(calendarId, it.id).execute()
        }
    }

    private fun findEvents(activity: Activity) =
        calendar.events().list(calendarId)
            .setQ(activity.searchString())
            .execute()
            .items
}
